ocdata/core/multi_adsorbate_slab_config.py

```diff
@@ -0,0 +1,192 @@
+from typing import List
+
+import numpy as np
+from ase import Atoms
+from ase.data import covalent_radii
+
+from ocdata.core import Adsorbate, AdsorbateSlabConfig, Slab
+
+
+class MultipleAdsorbateSlabConfig(AdsorbateSlabConfig):
+    """
+    Class to represent a slab with multiple adsorbates on it. This class only
+    returns a fixed combination of adsorbates placed on the surface. Unlike
+    AdsorbateSlabConfig which enumerates all possible adsorbate placements, this
+    problem gets combinatorially large.
+    """
+
+    def __init__(
+        self,
+        slab: Slab,
+        adsorbates: List[Adsorbate],
+        num_sites: int = 100,
+        num_configurations: int = 1,
+        interstitial_gap: float = 0.1,
+        mode: str = "random_site_heuristic_placement",
+    ):
+        assert mode in ["random", "heuristic", "random_site_heuristic_placement"]
+        assert interstitial_gap < 5 and interstitial_gap >= 0
+
+        self.slab = slab
+        self.adsorbates = adsorbates
+        self.num_sites = num_sites
+        self.interstitial_gap = interstitial_gap
+        self.mode = mode
+
+        self.sites = self.get_binding_sites(num_sites)
+        self.atoms_list, self.metadata_list = self.place_adsorbates_on_sites(
+            self.sites,
+            num_configurations,
+            interstitial_gap,
+        )
+
+    def place_adsorbates_on_sites(
+        self,
+        sites: list,
+        num_configurations: int = 1,
+        interstitial_gap: float = 0.1,
+    ):
+        """
+        Place the adsorbate at the given binding sites.
+
+        This method generates a fixed number of configurations where sites are
+        selected to ensure that adsorbate binding indices are at least a fair
+        distance away from each other (covalent radii + interstitial gap).
+        While this helps prevent adsorbate overlap it does not gaurantee it
+        since non-binding adsorbate atoms can overlap if the right combination
+        of angles is sampled.
+        """
+        # Build a fake atoms object with the positions as the sites.
+        # This allows us to easily compute distances while accounting for periodicity.
+        pseudo_atoms = Atoms(
+            [1] * len(sites), positions=sites, cell=self.slab.atoms.get_cell(), pbc=True
+        )
+        num_sites = len(sites)
+
+        atoms_list = []
+        metadata_list = []
+        ### NOTE: We can hard enforce these configurations to be non-overlapping.
```

Comment:
Any reason we switched to `###` comments? :)

Reply:
No reason, didn't realize the style we used previously. Updated :)